src/emqx_shared_sub.erl

```diff
@@ -264,23 +267,31 @@ pick(sticky, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
             {fresh, Sub0};
         false ->
             %% randomly pick one for the first message
-            {Type, Sub} = do_pick(random, ClientId, SourceTopic, Group, Topic, [Sub0 | FailedSubs]),
-            %% stick to whatever pick result
-            erlang:put({shared_sub_sticky, Group, Topic}, Sub),
-            {Type, Sub}
+            FailedSubs1 = maps_put_new(FailedSubs, Sub0, inactive),
+						case do_pick(random, ClientId, SourceTopic, Group, Topic, FailedSubs1) of
+              false -> false;
+              {Type, Sub} ->
+                %% stick to whatever pick result
+                erlang:put({shared_sub_sticky, Group, Topic}, Sub),
+                {Type, Sub}
+            end
     end;
 pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
     do_pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs).
 
 do_pick(Strategy, ClientId, SourceTopic, Group, Topic, FailedSubs) ->
     All = subscribers(Group, Topic),
-    case All -- FailedSubs of
+    case lists:filter(fun(Sub) -> not maps:is_key(Sub, FailedSubs) end, All) of
         [] when All =:= [] ->
             %% Genuinely no subscriber
             false;
         [] ->
             %% All offline? pick one anyway
-            {retry, pick_subscriber(Group, Topic, Strategy, ClientId, SourceTopic, All)};
+            %% We redispatch only to subs who dropped the message because inflight was full.
+            Dropped = maps_find_by(FailedSubs, fun({SubPid, FailReason}) ->
+                FailReason == dropped andalso is_alive_sub(SubPid)
+            end),
+            {retry, pick_subscriber(Group, Topic, Strategy, ClientId, SourceTopic, Dropped)};
```

Comment:
If I understood correctly, `maps_find_by` returns `error | {ok, SubPid, FailReason}`, and that is fed as the last argument to `pick_subscriber/6`, but `pick_subscriber/6` is expecting a list in that argument position. :thinking: